prov/shm/src/smr_ep.c

```diff
@@ -353,6 +353,47 @@ void smr_format_iov(struct smr_cmd *cmd, const struct iovec *iov, size_t count,
 	memcpy(cmd->msg.data.iov, iov, sizeof(*iov) * count);
 }
 
+int smr_format_ipc(struct smr_ep *ep, int64_t id, struct smr_cmd *cmd,
+		   const struct iovec *iov, enum fi_hmem_iface iface,
+		   uint64_t device, size_t total_len, struct smr_region *smr,
+		   struct smr_resp *resp, struct smr_tx_entry *pend)
+{
+	int ret;
+	void *base;
+
+	cmd->msg.hdr.op_src = smr_src_ipc;
+	cmd->msg.hdr.src_data = smr_get_offset(smr, resp);
+	cmd->msg.hdr.size = total_len;
+	cmd->msg.data.ipc_info.iface = iface;
+
+	if (iface == FI_HMEM_ZE) {
+		if (ep->sock_info->peers[id].state == SMR_CMAP_INIT)
+			smr_ep_exchange_fds(ep, id);
+		if (ep->sock_info->peers[id].state != SMR_CMAP_SUCCESS)
+			return -FI_EAGAIN;
+
+		ret = ofi_hmem_get_base_addr(iface, iov[0].iov_base, &base);
+		if (ret)
+			return ret;
+
+		ret = ze_hmem_get_shared_handle(ep->sock_info->my_fds[device],
+				base, &pend->fd,
+				(void **) &cmd->msg.data.ipc_info.fd_handle);
+		if (ret)
+			return ret;
+
+		cmd->msg.data.ipc_info.device = device;
+		cmd->msg.data.ipc_info.offset = (char *) iov[0].iov_base -
+						(char *) base;
+	} else {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"Interface not supported\n");
+		return -FI_EINVAL;
+	}
```

Comment:
You can move the check
```
if (iface != FI_HMEM_ZE)
    return -FI_EINVAL;
```
to the top of the function, which will improve readability.

Reply:
I left it this way anticipating that future interfaces would go through the same code path but I can change it if you prefer?

Reply:
The if path has a ze specific call in it.  If another device types goes through this path, you may end up with a switch/if statement immediately before the get_shared_handle call.  If that ends up being the case, then I would move the check to the top of the function, since you'll need some other check later.  I don't see an easy way to avoid checking things twice in this case without duplicating the rest of the code.

Regardless, this change can be later.

Reply:
Ahh good point. I'll update it now while I'm at it